File: algorithms/code/leetcode/lc299_bulls_and_cows/lc299_bulls_and_cows.py
#!/usr/bin/env python3
# https://leetcode.com/problems/bulls-and-cows/

import collections
import unittest


class Solution:
    def getHint(self, secret: str, guess: str) -> str:
        countKnown: collections.Counter = collections.Counter()
        countAsked: collections.Counter = collections.Counter()
        bulls = 0
        cows = 0
        i = 0
        while i < min(len(secret), len(guess)):
            # Count characters of each type
            known = secret[i]
            asked = guess[i]
            countKnown[known] += 1
            countAsked[asked] += 1
            # Check for a bull
            if known == asked:
                bulls += 1
            i += 1
        # Count the cows
        for asked in countAsked:
            if asked in countKnown:
                cows += min(countKnown[asked], countAsked[asked])
        # Remove the bulls from the cows
        cows -= bulls
        return '' + str(bulls) + 'A' + str(cows) + 'B'
        # It would have been faster to have one lookup table
        # and update cows up or down accordingly


class TestCode(unittest.TestCase):
    def test_1807__7810(self):
        self.assertEqual('1A3B', Solution().getHint('1807', '7810'))

    def test_1123__0111(self):
        self.assertEqual('1A1B', Solution().getHint('1123', '0111'))

    def test1122and2211(self):
        self.assertEqual('0A4B', Solution().getHint('1122', '2211'))

    def test11and10(self):
        self.assertEqual('1A0B', Solution().getHint('11', '10'))

File: algorithms/code/leetcode/lc299_bulls_and_cows/lc299_bulls_and_cows.rs
// https://leetcode.com/problems/bulls-and-cows/
// #medium

use std::collections::HashMap;

pub struct Solution;

impl Solution {
    pub fn get_hint(secret: String, guess: String) -> String {
        let mut count_known = HashMap::<char, i32>::new();
        let mut count_asked = HashMap::<char, i32>::new();
        let mut bulls = 0;
        let mut cows = 0;
        for i in 0..std::cmp::min(secret.chars().count(), guess.chars().count()) {
            // Count characters of each type
            let known = secret.chars().nth(i).unwrap();
            let asked = guess.chars().nth(i).unwrap();
            *count_known.entry(known).or_insert(0) += 1;
            *count_asked.entry(asked).or_insert(0) += 1;
            // Check for a bull
            if known == asked {
                bulls += 1;
            }
        }
        // Count the cows
        for asked in count_asked.keys() {
            if count_known.contains_key(asked) {
                cows += std::cmp::min(
                    count_known.get(asked).unwrap(),
                    count_asked.get(asked).unwrap(),
                );
            }
        }
        // Remove the bulls from the cows
        cows -= bulls;
        format!("{bulls}A{cows}B")
        // It would have been faster to have one lookup table and update cows up or down
        // accordingly
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_1807_and_7810() {
        assert_eq!(
            Solution::get_hint("1807".to_string(), "7810".to_string()),
            "1A3B",
        );
    }

    #[test]
    fn test_1123_and_0111() {
        assert_eq!(
            Solution::get_hint("1123".to_string(), "0111".to_string()),
            "1A1B",
        );
    }

    #[test]
    fn test_1122_and_2211() {
        assert_eq!(
            Solution::get_hint("1122".to_string(), "2211".to_string()),
            "0A4B",
        );
    }

    #[test]
    fn test_11_and_10() {
        assert_eq!(
            Solution::get_hint("11".to_string(), "10".to_string()),
            "1A0B",
        );
    }
}
